<div>
    <h1>Cars Project data processing</h1>
    <div>This project corresponds to pre-process the data (study duplicates, missing values, try to restore the missing values) and some exploratory   data analysis of the cars dealer data frame, in order to develop a web application and deploy it in Render</div>
</div>

In [1]:
from scipy import stats as st
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt

In [2]:
df = pd.read_csv("C://Users/laura/vehicles_us.csv")
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    47906 non-null  float64
 2   model         51525 non-null  object 
 3   condition     51525 non-null  object 
 4   cylinders     46265 non-null  float64
 5   fuel          51525 non-null  object 
 6   odometer      43633 non-null  float64
 7   transmission  51525 non-null  object 
 8   type          51525 non-null  object 
 9   paint_color   42258 non-null  object 
 10  is_4wd        25572 non-null  float64
 11  date_posted   51525 non-null  object 
 12  days_listed   51525 non-null  int64  
dtypes: float64(4), int64(2), object(7)
memory usage: 5.1+ MB



<div class="alert alert-info">
<div>When reading the dataframe, a document with 13 columns and 51525 entries is found. The columns correspond to the information of each of the vehicles available in the car store and include the following data: the price of the vehicle, the year, the model, the vehicle code, the number of cylinders, the type of fuel used. used, the odometer, the type of transmission, the type of vehicle, the color of the exterior paint, the day it was published for sale and the number of days it has been on sale.</div>

In [3]:
print(df.isna().sum())

price               0
model_year       3619
model               0
condition           0
cylinders        5260
fuel                0
odometer         7892
transmission        0
type                0
paint_color      9267
is_4wd          25953
date_posted         0
days_listed         0
dtype: int64


<div class="alert alert-info">
<div>When searching for missing values ​​in each of the columns, the following were found: 3619 missing values ​​in the model year corresponding to 7%, 5260 missing values ​​in the number of cylinders corresponding to 10%, 7892 missing values ​​in the odometer corresponding to 15%. , 9267 missing values ​​in the paint information corresponding to 17% and 25953, in is_4wd corresponding to 50%</div>


In [50]:
model_year_mean=df['model_year'].mean()
df['model_year']=df['model_year'].fillna(model_year_mean)

print(model_year_mean)

print(df.isna().sum())


2009.75046966977
price           0
model_year      0
model           0
condition       0
cylinders       0
fuel            0
odometer        0
transmission    0
type            0
paint_color     0
is_4wd          0
date_posted     0
days_listed     0
dtype: int64


<div class="alert alert-info">
<div>
    To fix the year column of the vehicle model, this is quantitative data whose values ​​are from 1908 to 2019. The average of the values ​​is: 2009.75046966977 and the median is 2010. So it is decided to fill in the spaces blank with the value 2010 representing the values ​​in the column
</div>

In [64]:
cylinders_mean=df['cylinders'].median()
df['cylinders']=df['cylinders'].fillna(cylinders_mean)

print(cylinders_mean)

print(df.isna().sum())

6.0
price           0
model_year      0
model           0
condition       0
cylinders       0
fuel            0
odometer        0
transmission    0
type            0
paint_color     0
is_4wd          0
date_posted     0
days_listed     0
dtype: int64


<div class="alert alert-info">
<div>
    To fix the column of the number of cylinders, this is a quantitative data whose values ​​are from 3 to 12. The average of the values ​​is: 6.125235058899817 and the median is 6.0. Therefore, it is decided to fill the blank spaces with the value 6.0, which represents the values ​​in the column.
</div>

In [75]:
odometer_median=df['odometer'].median()
df['odometer']=df['odometer'].fillna(odometer_median)

print(odometer_median)

print(df.isna().sum())


113000.0
price           0
model_year      0
model           0
condition       0
cylinders       0
fuel            0
odometer        0
transmission    0
type            0
paint_color     0
is_4wd          0
date_posted     0
days_listed     0
dtype: int64


<div class="alert alert-info">
<div>
    To fix the odometer value column, this is a quantitative data whose values ​​are from 0 to 920000. The average of the values ​​is: 115162.3521785541 and the median is 113000.0. So it is decided to fill the blank spaces with the value 113000.0 since data has significant outliers
</div>

In [ ]:
df['paint_color']=df['paint_color'].fillna('unknown')


<div class="alert alert-info">
<div>
The vehicle's external paint column corresponds to a categorical value, so the fillna method is used to replace the empty cells with the "unknown" value.
</div>

In [ ]:
df['is_4wd']=df['is_4wd'].fillna(1)
print(df.isna().sum())

<div class="alert alert-info">
<div>
The is_4wd column of the vehicle corresponds to a categorical value, so the fillna method is used to replace the empty cells with the value 1
</div>

In [36]:
print(df.duplicated().sum())

0


<div class="alert alert-info">
<div>
No duplicate values ​​were found in the dataframe
</div>

In [84]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    51525 non-null  float64
 2   model         51525 non-null  object 
 3   condition     51525 non-null  object 
 4   cylinders     51525 non-null  float64
 5   fuel          51525 non-null  object 
 6   odometer      51525 non-null  float64
 7   transmission  51525 non-null  object 
 8   type          51525 non-null  object 
 9   paint_color   51525 non-null  object 
 10  is_4wd        51525 non-null  float64
 11  date_posted   51525 non-null  object 
 12  days_listed   51525 non-null  int64  
dtypes: float64(4), int64(2), object(7)
memory usage: 5.1+ MB


In [114]:
df['model_year']=df['model_year'].astype('int64')
df['cylinders']=df['cylinders'].astype('int64')
df['date_posted']=pd.to_datetime(df['date_posted'],format='%Y-%m-%d')
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   price         51525 non-null  int64         
 1   model_year    51525 non-null  int64         
 2   model         51525 non-null  object        
 3   condition     51525 non-null  object        
 4   cylinders     51525 non-null  int64         
 5   fuel          51525 non-null  object        
 6   odometer      51525 non-null  float64       
 7   transmission  51525 non-null  object        
 8   type          51525 non-null  object        
 9   paint_color   51525 non-null  object        
 10  is_4wd        51525 non-null  float64       
 11  date_posted   51525 non-null  datetime64[ns]
 12  days_listed   51525 non-null  int64         
dtypes: datetime64[ns](1), float64(2), int64(4), object(6)
memory usage: 5.1+ MB


price  model_year           model  condition  cylinders fuel  odometer  \
0   9400        2011          bmw x5       good          6  gas  145000.0   
1  25500        2009      ford f-150       good          6  gas   88705.0   
2   5500        2013  hyundai sonata   like new          4  gas  110000.0   
3   1500        2003      ford f-150       fair          8  gas  113000.0   
4  14900        2017    chrysler 200  excellent          4  gas   80903.0   

  transmission    type paint_color  is_4wd date_posted  days_listed  
0    automatic     SUV     unknown     1.0  2018-06-23           19  
1    automatic  pickup       white     1.0  2018-10-19           50  
2    automatic   sedan         red     1.0  2019-02-07           79  
3    automatic  pickup     unknown     1.0  2019-03-22            9  
4    automatic   sedan       black     1.0  2019-04-02           28

<div class="alert alert-info">
<div>
    In order to fix the data type with the concept corresponding to each column, the following columns will be changed
</div>
<li>The vehicle model year corresponds to a type int 64</li>
<li>The cylinders number corresponds to a type int 64</li>
<li>The date posted corresponds to a type datetime</li>



<div class="alert alert-info">
<h2>Conclusion</h2>
 <div>
    The dataframe corresponds to a large amount of data of which the is_4wd column was the one with the most empty data with 50% of the data value, however the remaining 50% corresponds to the value of 1 so a value of 1 was inferred. 1 for missing data.
The exterior paint column does not have a major impact on the data analysis so filling in missing values ​​with the word "unknown" is appropriate.
For values ​​such as the model year, the number of cylinders and the odometer, basic statistics were applied to complete the missing values ​​with the mean or median since these correspond to quantitative data.

</div>